### ISSUE TITLE
Branched discussion from kubernetes_asyncio switch - is ThreadPool workaround needed?

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/kubespawner#567]
# TL;DR

I considered if a workaround is still needed, also after `kubernetes_asyncio` is used to replace `kubernetes` in #563, and concluded yes it is.

### Background

I'm reviewing #563 that replaces `kubernetes` with `kubernetes_asyncio`. Doing that, I'm considering if we should remove a workaround or not.

### Regarding this workaround

This is a workaround introduced in #169.

https://github.com/jupyterhub/kubespawner/blob/a6698161957c1018b879fafe3ff71c2dd5c35d79/kubespawner/clients.py#L12-L19

### Do we still need it?

A ThreadPool object was in the past initialized by the kubernetes client without a specified number of threads to be used, which led to a number of threads matching the number of CPUs. Since `kubernetes>=9` they initialize the ThreadPool object with `1` thread by default.

[![image](https://user-images.githubusercontent.com/3837114/154517336-ef062c51-d56d-428d-ad4e-bbfca2241614.png)](https://github.com/kubernetes-client/python/commit/b4d11b02a3479e63957a729614a616002f13e9c4#diff-59aff6ce4d28aa662f8b411b9d0dfe4f3e949c32a5edaf8e08905b58e7a41ee3L67-R69)

Does this make us no longer need the patch, or is it problematic that a ThreadPool object is instantiated at all? I think so, but **does this change with switching to `kubernetes_asyncio`**?

### How does the hack work?

https://github.com/jupyterhub/kubespawner/blob/a6698161957c1018b879fafe3ff71c2dd5c35d79/kubespawner/clients.py#L18-L19

This modifies [the following code](https://github.com/tomplus/kubernetes_asyncio/blob/c88b3b9c78a9388ef3af00858b8c516b84b0bf30/kubernetes_asyncio/client/api_client.py#L99-L107):

```python
    @property
    def pool(self):
        """Create thread pool on first request
         avoids instantiating unused threadpool for blocking clients.
        """
        if self._pool is None:
            atexit.register(self.close)
            self._pool = ThreadPool(self.pool_threads)
        return self._pool
```

From this, I presume that when `self.pool` is referenced, a [`Mock()`](https://docs.python.org/3/library/unittest.mock.html#the-mock-class) instance is returned, which just happily accepts everything but does nothing.

### Will `kubernetes_asyncio` use the `pool`?

Depends, is `async_req` passed when [`ApiClient`](https://github.com/tomplus/kubernetes_asyncio/blob/HEAD/kubernetes_asyncio/client/api_client.py#L33) instances calls `call_api`?

```python
    def call_api(self, resource_path, ...
        if not async_req:
            return self.__call_api(resource_path, ...
        return self.pool.apply_async(self.__call_api, (resource_path, ...
```

### Resolution

I think neither `kubernetes` or `kubernetes_asyncion` makes api requests by passing `async_req` parameters. But I note there is a difference between a `__call_api` function.

- [kubernetes](https://github.com/kubernetes-client/python/blob/ada2a7d73544e5be6a266c326f411b43c993e973/kubernetes/client/api_client.py#L120) defines `__call_api` without `async`.
- [kubernetes_asyncio](https://github.com/tomplus/kubernetes_asyncio/blob/c88b3b9c78a9388ef3af00858b8c516b84b0bf30/kubernetes_asyncio/client/api_client.py#L121) defines `__call_api` with `async`.

Overall, I conclude we shouldn't have a threadpool, and Mocking it to never initialize a ThreadPool with one thread is fine - because it will never be referenced no matter what.

**Comment by athornton** [comment]
I came to basically the same conclusion.  I don't _think_ we need it anymore, but it's super-lightweight and it didn't feel like there was any harm in leaving it around.  We should not need a threadpool since we're no longer using Threading at all.
